### PR TITLE
Adopt typed throws in withUnsafe(Mutable)BufferPointer

### PR DESCRIPTION
**Pull request by DougGregor**
Fixes rdar://131405937

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

File: stdlib/public/core/ArrayBuffer.swift [hunk 1]
```diff
@@ -567,29 +567,56 @@ extension _ArrayBuffer {
     }
   }
 
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @inlinable
```
Comment (slavapestov) [inline stdlib/public/core/ArrayBuffer.swift:572]: This should be @usableFromInline, not @inlinable.

File: stdlib/public/core/Array.swift [hunk 2]
```diff
@@ -1595,56 +1605,24 @@ extension Array {
...
-  ///    If no such storage exists, it is created. If `body` has a return value, that value is also
-  ///   used as the return value for the `withUnsafeMutableBufferPointer(_:)`
-  ///   method. The pointer argument is valid only for the duration of the
-  ///   method's execution.
-  /// - Returns: The return value, if any, of the `body` closure parameter.
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
   @_semantics("array.withUnsafeMutableBufferPointer")
   @_effects(notEscaping self.value**)
   @inlinable // FIXME(inline-always)
```
Comment (slavapestov) [inline stdlib/public/core/Array.swift:1619]: This should be @usableFromInline, not @inlinable.

File: stdlib/public/core/ArrayBuffer.swift [hunk 3]
```diff
@@ -567,29 +567,56 @@ extension _ArrayBuffer {
...
+    _ body: (UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    return try withUnsafeMutableBufferPointer(body)
+  }
+
   /// Call `body(p)`, where `p` is an `UnsafeMutableBufferPointer`
   /// over the underlying contiguous storage.
   ///
   /// - Precondition: Such contiguous storage exists or the buffer is empty.
   @inlinable
```
Comment (slavapestov) [inline stdlib/public/core/ArrayBuffer.swift:615]: Nit: @aeic implies @inlinable

File: stdlib/public/core/ArraySlice.swift [hunk 4]
```diff
@@ -1194,12 +1204,28 @@ extension ArraySlice {
...
+  public func withUnsafeBufferPointer<R, E>(
+    _ body: (UnsafeBufferPointer<Element>) throws(E) -> R
+  ) throws(E) -> R {
     return try _buffer.withUnsafeBufferPointer(body)
   }
 
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @_semantics("array.withUnsafeMutableBufferPointer")
+  @inlinable // FIXME(inline-always)
```
Comment (slavapestov) [inline stdlib/public/core/ArraySlice.swift:1217]: @usableFromInline not @inlinable

File: stdlib/public/core/ContiguousArray.swift [hunk 5]
```diff
@@ -1135,12 +1145,28 @@ extension ContiguousArray {
...
+  public func withUnsafeBufferPointer<R, E>(
+    _ body: (UnsafeBufferPointer<Element>) throws(E) -> R
+  ) throws(E) -> R {
     return try _buffer.withUnsafeBufferPointer(body)
   }
 
+  // Superseded by the typed-throws version of this function, but retained
+  // for ABI reasons.
+  @_semantics("array.withUnsafeMutableBufferPointer")
+  @inlinable // FIXME(inline-always)
```
Comment (slavapestov) [inline stdlib/public/core/ContiguousArray.swift:1158]: Same

**Comment by DougGregor** [comment]
Thank you, @slavapestov !

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test

File: stdlib/public/core/Array.swift [hunk 6]
```diff
@@ -1594,59 +1604,54 @@ extension Array {
...
-  ///   appear to have its correct value. Instead, use only the
-  ///   `UnsafeMutableBufferPointer` argument to `body`.
-  ///
-  /// - Parameter body: A closure with an `UnsafeMutableBufferPointer`
-  ///   parameter that points to the contiguous storage for the array.
-  ///    If no such storage exists, it is created. If `body` has a return value, that value is also
-  ///   used as the return value for the `withUnsafeMutableBufferPointer(_:)`
-  ///   method. The pointer argument is valid only for the duration of the
-  ///   method's execution.
-  /// - Returns: The return value, if any, of the `body` closure parameter.
```
Comment (glessard) [inline stdlib/public/core/Array.swift:1639]: This doc-comment should remain on the redone public method.

File: test/api-digester/stability-stdlib-source.swift [hunk 7]
```diff
@@ -4,7 +4,7 @@
 // This currently fails on non-Intel architectures due to no baseline being
 // available and it is not possible to filter across architectures in the
 // output.
-// XFAIL: CPU=arm64 || CPU=arm64e
```
Comment (glessard) [inline test/api-digester/stability-stdlib-source.swift:7]: Nice!

**Review by glessard — APPROVED** [review]

File: stdlib/public/core/Array.swift (hunk 6 above)
Comment (DougGregor) [reply stdlib/public/core/Array.swift:1639]: Thank you for catching this! I'm pushing a fix now.

**Comment by DougGregor** [comment]
@swift-ci please test

**Review by stephentyrone — APPROVED** [review]

**Comment by DougGregor** [comment]
@swift-ci please test

**Pull request merged**